File: src/fetch_data/trending_scraper.py
"""Scrape trending repositories from GitHub without API token"""
import logging
import requests
from datetime import datetime
from typing import List, Dict, Any
from bs4 import BeautifulSoup
from sqlalchemy.orm import Session
from src.database.models import Project, TrendingSnapshot

logger = logging.getLogger(__name__)


class TrendingScraper:
    """Scrape GitHub trending page without API token"""

    BASE_URL = "https://github.com/trending"

    def __init__(self, db_session: Session):
        """
        Initialize trending scraper

        Args:
            db_session: Database session
        """
        self.db = db_session
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        })

    def scrape_trending(self, language: str = None, since: str = "daily") -> List[Dict[str, Any]]:
        """
        Scrape trending repositories from GitHub

        Args:
            language: Programming language filter (e.g., "python")
            since: Time range - "daily", "weekly", or "monthly"

        Returns:
            List of repository data dictionaries
        """
        # Build URL
        url = self.BASE_URL
        params = {}
        if language:
            params['spoken_language_code'] = language
        if since:
            params['since'] = since

        logger.info(f"Scraping GitHub trending: {url} with params {params}")

        try:
            response = self.session.get(url, params=params, timeout=30)
            response.raise_for_status()

            soup = BeautifulSoup(response.text, 'lxml')

            # Find all repository articles
            repos = soup.select('article.Box-row')

            trending_data = []
            for rank, repo in enumerate(repos, 1):
                try:
                    # Extract repository name and owner
                    h2 = repo.select_one('h2 a')
                    if not h2:
                        continue

                    full_name = h2['href'].strip('/')
                    name = full_name.split('/')[-1]
                    owner = full_name.split('/')[0]

                    # Extract description
                    desc_elem = repo.select_one('p')
                    description = desc_elem.text.strip() if desc_elem else ""

                    # Extract language
                    lang_elem = repo.select_one('[itemprop="programmingLanguage"]')
                    language = lang_elem.text.strip() if lang_elem else None

                    # Extract stars
                    stars_elem = repo.select_one('a[href*="/stargazers"]')
                    stars = 0
                    if stars_elem:
                        stars_text = stars_elem.text.strip().replace(',', '')
                        try:
                            stars = int(stars_text)
                        except ValueError:
                            pass

                    # Build URL
                    url = f"https://github.com/{full_name}"

                    repo_data = {
                        'rank': rank,
                        'name': name,
                        'full_name': full_name,
                        'owner': owner,
                        'description': description,
                        'language': language,
                        'stars': stars,
                        'url': url,
                    }

                    trending_data.append(repo_data)

                except Exception as e:
                    logger.warning(f"Failed to parse repo: {e}")
                    continue

            logger.info(f"Successfully scraped {len(trending_data)} trending repositories")
            return trending_data

        except Exception as e:
            logger.error(f"Error scraping trending: {e}")
            raise

    def save_to_database(self, trending_data: List[Dict[str, Any]]) -> int:
        """
        Save scraped data to database

        Args:
            trending_data: List of repository data dictionaries

        Returns:
            Number of saved repositories
        """
        saved_count = 0
        snapshot_date = datetime.now()

        try:
            for repo_data in trending_data:
                # Check if project exists
                project = self.db.query(Project).filter(
                    Project.full_name == repo_data['full_name']
                ).first()

                if project:
                    # Update existing project
                    project.stars = repo_data['stars']
                    project.description = repo_data['description']
                    project.updated_at = datetime.now()
                else:
                    # Create new project
                    project = Project(
                        name=repo_data['name'],
                        full_name=repo_data['full_name'],
                        description=repo_data['description'],
                        language=repo_data['language'],
                        stars=repo_data['stars'],
                        url=repo_data['url'],
                        created_at=datetime.now()
                    )
                    self.db.add(project)
                    self.db.flush()

                # Create snapshot
                snapshot = TrendingSnapshot(
                    date=snapshot_date,
                    project_id=project.id,
                    stars_at_snapshot=repo_data['stars'],
                    rank=repo_data['rank'],
                    created_at=datetime.now()
                )
                self.db.add(snapshot)
                saved_count += 1

            self.db.commit()
            logger.info(f"Saved {saved_count} repositories to database")
            return saved_count

        except Exception as e:
            self.db.rollback()
            logger.error(f"Error saving to database: {e}")
            raise

    def fetch_and_save(self, language: str = None, since: str = "daily") -> int:
        """
        Scrape and save trending repositories

        Args:
            language: Programming language filter
            since: Time range

        Returns:
            Number of saved repositories
        """
        trending_data = self.scrape_trending(language=language, since=since)
        return self.save_to_database(trending_data)

File: src/fetch/TrendingScraper.ts
import axios from 'axios';
import * as cheerio from 'cheerio';
import { Repository } from 'typeorm';
import { Project } from '../database/models/Project';
import { TrendingSnapshot } from '../database/models/TrendingSnapshot';

export interface TrendingRepoData {
  rank: number;
  name: string;
  full_name: string;
  owner: string;
  description: string;
  language?: string;
  stars: number;
  url: string;
}

export class TrendingScraper {
  private static readonly BASE_URL = 'https://github.com/trending';
  private projectRepo: Repository<Project>;
  private snapshotRepo: Repository<TrendingSnapshot>;

  constructor(
    projectRepo: Repository<Project>,
    snapshotRepo: Repository<TrendingSnapshot>
  ) {
    this.projectRepo = projectRepo;
    this.snapshotRepo = snapshotRepo;
  }

  async scrapeTrending(
    language?: string,
    since: 'daily' | 'weekly' | 'monthly' = 'daily'
  ): Promise<TrendingRepoData[]> {
    const url = TrendingScraper.BASE_URL;
    const params: Record<string, string> = {};

    if (language) {
      params.spoken_language_code = language;
    }
    if (since) {
      params.since = since;
    }

    console.log(`Scraping GitHub trending: ${url}`, params);

    try {
      const response = await axios.get(url, {
        params,
        headers: {
          'User-Agent':
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
          'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
          'Accept-Language': 'en-US,en;q=0.5',
          'Accept-Encoding': 'gzip, deflate, br',
          'Connection': 'keep-alive',
          'Upgrade-Insecure-Requests': '1',
        },
        timeout: 30000,
        maxRedirects: 5,
      });

      const $ = cheerio.load(response.data);
      const repos = $('article.Box-row');
      const trendingData: TrendingRepoData[] = [];

      repos.each((index, element) => {
        try {
          const $repo = $(element);

          // Extract repository name and owner
          const h2 = $repo.find('h2 a');
          if (!h2.length) return;

          const href = h2.attr('href');
          if (!href) return;

          const fullName = href.replace(/^\//, '');
          const parts = fullName.split('/');
          const owner = parts[0];
          const name = parts[1];

          // Extract description
          const descElem = $repo.find('p');
          const description = descElem.text().trim();

          // Extract language
          const langElem = $repo.find('[itemprop="programmingLanguage"]');
          const language = langElem.length ? langElem.text().trim() : undefined;

          // Extract stars
          let stars = 0;
          const starsElem = $repo.find('a[href*="/stargazers"]');
          if (starsElem.length) {
            const starsText = starsElem.text().trim().replace(/,/g, '');
            const parsedStars = parseInt(starsText, 10);
            if (!isNaN(parsedStars)) {
              stars = parsedStars;
            }
          }

          const url = `https://github.com/${fullName}`;
          const rank = index + 1;

          trendingData.push({
            rank,
            name,
            full_name: fullName,
            owner,
            description,
            language,
            stars,
            url,
          });
        } catch (error) {
          console.warn(`Failed to parse repo:`, error);
        }
      });

      console.log(
        `Successfully scraped ${trendingData.length} trending repositories`
      );
      return trendingData;
    } catch (error) {
      console.error('Error scraping trending:', error);
      throw error;
    }
  }

  async saveToDatabase(trendingData: TrendingRepoData[]): Promise<number> {
    let savedCount = 0;
    const snapshotDate = new Date();

    try {
      for (const repoData of trendingData) {
        // Check if project exists
        let project = await this.projectRepo.findOne({
          where: { full_name: repoData.full_name },
        });

        if (project) {
          // Update existing project
          project.stars = repoData.stars;
          project.description = repoData.description;
          project.updated_at = new Date();
          await this.projectRepo.save(project);
        } else {
          // Create new project
          project = this.projectRepo.create({
            name: repoData.name,
            full_name: repoData.full_name,
            description: repoData.description,
            language: repoData.language,
            stars: repoData.stars,
            url: repoData.url,
          });
          await this.projectRepo.save(project);
        }

        // Create snapshot
        const snapshot = this.snapshotRepo.create({
          date: snapshotDate,
          project_id: project.id,
          stars_at_snapshot: repoData.stars,
          rank: repoData.rank,
        });
        await this.snapshotRepo.save(snapshot);
        savedCount++;
      }

      console.log(`Saved ${savedCount} repositories to database`);
      return savedCount;
    } catch (error) {
      console.error('Error saving to database:', error);
      throw error;
    }
  }

  async fetchAndSave(
    language?: string,
    since: 'daily' | 'weekly' | 'monthly' = 'daily'
  ): Promise<number> {
    const trendingData = await this.scrapeTrending(language, since);
    return await this.saveToDatabase(trendingData);
  }
}
